File: challenge_2/go/makernaren/single_number.go
// Take an array of mixed data types and find the element that occurs
// only once in it.
package main

import (
	"fmt"
)

// Finds the single occurance, takes in interface as input and output params
//as elements in the given array may be of any data type
func FindSingle(input []interface{}) (result interface{}) {
	// Create a hashmap, where element of that array is key and its number of
	// occurances is the value.
	occurances := make(map[interface{}]int)
	for _, key := range input {
		occurances[key]++
	}
	// Return when we find the single occurance.
	for k, v := range occurances {
		if v == 1 {
			result = k
			break
		}
	}
	return result
}

func main() {
	given := []interface{}{2, "a", "l", 3, "l", 4, "k", 2, 3, 4, "a", 6, "c",
		4, "m", 6, "m", "k", 9, 10, 9, 8, 7, 8, 10, 7}
	fmt.Printf("Given array is : %v \n", given)
	result := FindSingle(given)
	fmt.Printf("Element that occur only once : %v \n", result)
}

File: challenge_2/rust/makernaren/src/single_number.rs
use std::collections::HashMap;

// Finds the unique character in given array
pub fn find_single(input: Vec<i32>) -> i32 {
	// Create hashmap, let the key be the element in input array
	// and value be its count. Then return the key with count of 1.
    let mut occurances: HashMap<i32, i32> = HashMap::new();
    for key in input {
    *occurances.entry(key).or_insert(0) += 1;
    }
   let mut result: i32 = 0;
    for (key, val) in occurances.iter() {
        if *val == 1 {
            result = key.to_owned();
        }
    }
    return result
}


#[cfg(test)]
mod tests {
	use super::*;
	#[test] 
	fn test_int() {
		let expected = 5;
		let input = vec![2, 3, 4, 2, 3, 5, 4, 6, 4, 6, 9, 10,9 ,8 ,7 ,8 ,10 ,7];
		assert_eq!(expected.to_owned(), find_single(input));	
	}
}